Make the derivative of masked_fill more efficient

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83515

There's no need to add all the zeros if we extract all the non-zero
elements.

lezcano committed:
Make the derivative of masked_fill more efficient

There's no need to add all the zeros if we extract all the non-zero
elements.

[ghstack-poisoned]

Lezcano commented:
@albanD @soulitzer could you have a look?

lezcano committed:
Update on "Make the derivative of masked_fill more efficient"


There's no need to add all the zeros if we extract all the non-zero
elements.

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -497,6 +497,13 @@ Tensor sgn_backward(const Tensor& x, const Tensor& gx, const Tensor& sgn) {
+  // masked_select does not work well with functorch, as its shape is data-dependent
+  return areAnyTensorSubclassLike({grad, mask}) ?
+          at::where(mask, grad, 0).sum() :
+          grad.masked_select(mask).sum();
```
That looks composite compliant to me. Why the if?

Lezcano replied:
As the comment says, the shape of the output of `masked_select` is data dependent and functorch errors out on this. Now, in my opinion, the batching rule for `masked_select` should not return a batch of vectors of different size, but a long vector (perhaps with batching size of 1 if functorch requires it). cc @zou3519

Lezcano replied:
fwiw, this is the error: https://github.com/pytorch/pytorch/runs/7859535004?check_suite_focus=true

Lezcano replied:
fwiw, this is the error: https://github.com/pytorch/pytorch/runs/7859535004?check_suite_focus=true

zou3519 replied:
> Now, in my opinion, the batching rule for masked_select should not return a batch of vectors of different size, but a long vector (perhaps with batching size of 1 if functorch requires it)

`vmap(f)(xs)` has semantics equivalent to the following: in the absence of side-effects, it is equivalent to running f in a for-loop on each x and stack-ing the results.

Because masked_select is guaranteed to return a 1D tensor, then vmap(masked_select) should return a 2D tensor

Lezcano replied:
Makes sense, thanks!

zou3519 replied:
minor nit for the comment: replace "functorch" with "vmap". The other functorch transforms don't have problems with masked_select, it's just vmap.

albanD reviewed (approved):
Well I should read the comments in more details :p

lezcano committed:
Update on "Make the derivative of masked_fill more efficient"


There's no need to add all the zeros if we extract all the non-zero
elements.

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -g

lezcano committed:
Update on "Make the derivative of masked_fill more efficient"


There's no need to add all the zeros if we extract all the non-zero
elements.

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -g
